### ISSUE TITLE
[receiver/kafkareceiver]: Pipeline seems to be blocked when receiver can't connect to brokers

### ISSUE THREAD

**Issue by antonblock** [open-telemetry/opentelemetry-collector-contrib#24548]
### Component(s)

receiver/kafka

## Description

When starting a collector with a Kafka receiver, the receiver seems to block other receivers in the pipeline from starting if it can't connect to the specified Kafka system. This can be for logs, metrics, or traces.

## Steps to Reproduce

1. Configure a Kafka receiver to collect from a system it can't reach, e.g. invalid host or a host not running Kafka
2. Configure the `hostmetrics` receiver to collect every 10 seconds
3. Create a pipeline consisting of both those receivers exporting with `logging`

## Expected Result

Kafka fails to collect, but `hostmetrics` is still successfully collecting and exporting.

## Actual Result

Kafka endlessly logs connection errors and the `hostmetrics` receiver doesn't seem to be running or getting exported.

### Collector version

v0.81.0

## Environment
OS: macOS
Compiler(if manually compiled): go 1.20.5

### OpenTelemetry Collector configuration

```yaml
receivers:
  kafka:
    brokers:
      - localhost:9092
    client_id: otel-collector
    group_id: otel-collector
    metadata:
      full: false
    protocol_version: 2.0.0
    topic: otlp_metrics
  hostmetrics:
    collection_interval: 10s
    scrapers:
      load:
      filesystem:
      memory:
exporters:
  otlphttp:
    endpoint: http://localhost:9123
  logging:
    verbosity: detailed
service:
  telemetry:
    logs:
      level: "debug"
  pipelines:
    metrics:
      receivers: [kafka,hostmetrics]
      exporters: [otlphttp,logging]
```

### Log output

```shell
2023-07-25T12:30:47.059-0400	info	service/telemetry.go:81	Setting up own telemetry...
2023-07-25T12:30:47.060-0400	info	service/telemetry.go:104	Serving Prometheus metrics	{"address": ":8888", "level": "Basic"}
2023-07-25T12:30:47.060-0400	info	exporter@v0.81.0/exporter.go:275	Development component. May change in the future.	{"kind": "exporter", "data_type": "metrics", "name": "logging"}
2023-07-25T12:30:47.061-0400	debug	exporter@v0.81.0/exporter.go:273	Stable component.	{"kind": "exporter", "data_type": "metrics", "name": "otlphttp"}
2023-07-25T12:30:47.061-0400	debug	receiver@v0.81.0/receiver.go:294	Beta component. May change in the future.	{"kind": "receiver", "name": "kafka", "data_type": "metrics"}
2023-07-25T12:30:47.062-0400	debug	receiver@v0.81.0/receiver.go:294	Beta component. May change in the future.	{"kind": "receiver", "name": "hostmetrics", "data_type": "metrics"}
2023-07-25T12:30:47.062-0400	info	service/service.go:131	Starting otelcontribcol...	{"Version": "0.81.0-dev", "NumCPU": 10}
2023-07-25T12:30:47.062-0400	info	extensions/extensions.go:30	Starting extensions...
2023-07-25T12:30:47.063-0400	info	kafkareceiver@v0.81.0/kafka_receiver.go:238	Waiting for consumer group to start	{"kind": "receiver", "name": "kafka", "data_type": "metrics"}
2023-07-25T12:30:47.826-0400	error	kafkareceiver@v0.81.0/kafka_receiver.go:250	Error from consumer	{"kind": "receiver", "name": "kafka", "data_type": "metrics", "error": "kafka: client has run out of available brokers to talk to: dial tcp [::1]:9092: connect: connection refused"}
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver.(*kafkaMetricsConsumer).consumeLoop
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver@v0.81.0/kafka_receiver.go:250
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver.(*kafkaMetricsConsumer).Start.func1
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver@v0.81.0/kafka_receiver.go:234
2023-07-25T12:30:48.590-0400	error	kafkareceiver@v0.81.0/kafka_receiver.go:250	Error from consumer	{"kind": "receiver", "name": "kafka", "data_type": "metrics", "error": "kafka: client has run out of available brokers to talk to: dial tcp [::1]:9092: connect: connection refused"}
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver.(*kafkaMetricsConsumer).consumeLoop
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver@v0.81.0/kafka_receiver.go:250
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver.(*kafkaMetricsConsumer).Start.func1
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver@v0.81.0/kafka_receiver.go:234
2023-07-25T12:30:49.368-0400	error	kafkareceiver@v0.81.0/kafka_receiver.go:250	Error from consumer	{"kind": "receiver", "name": "kafka", "data_type": "metrics", "error": "kafka: client has run out of available brokers to talk to: dial tcp [::1]:9092: connect: connection refused"}
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver.(*kafkaMetricsConsumer).consumeLoop
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver@v0.81.0/kafka_receiver.go:250
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver.(*kafkaMetricsConsumer).Start.func1
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver@v0.81.0/kafka_receiver.go:234
2023-07-25T12:30:50.137-0400	error	kafkareceiver@v0.81.0/kafka_receiver.go:250	Error from consumer	{"kind": "receiver", "name": "kafka", "data_type": "metrics", "error": "kafka: client has run out of available brokers to talk to: dial tcp [::1]:9092: connect: connection refused"}
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver.(*kafkaMetricsConsumer).consumeLoop
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver@v0.81.0/kafka_receiver.go:250
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver.(*kafkaMetricsConsumer).Start.func1
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver@v0.81.0/kafka_receiver.go:234
2023-07-25T12:30:50.898-0400	error	kafkareceiver@v0.81.0/kafka_receiver.go:250	Error from consumer	{"kind": "receiver", "name": "kafka", "data_type": "metrics", "error": "kafka: client has run out of available brokers to talk to: dial tcp [::1]:9092: connect: connection refused"}
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver.(*kafkaMetricsConsumer).consumeLoop
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver@v0.81.0/kafka_receiver.go:250
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver.(*kafkaMetricsConsumer).Start.func1
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver@v0.81.0/kafka_receiver.go:234
2023-07-25T12:30:51.661-0400	error	kafkareceiver@v0.81.0/kafka_receiver.go:250	Error from consumer	{"kind": "receiver", "name": "kafka", "data_type": "metrics", "error": "kafka: client has run out of available brokers to talk to: dial tcp [::1]:9092: connect: connection refused"}
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver.(*kafkaMetricsConsumer).consumeLoop
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver@v0.81.0/kafka_receiver.go:250
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver.(*kafkaMetricsConsumer).Start.func1
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkareceiver@v0.81.0/kafka_receiver.go:234

** above errors are repeated indefinitely, with no other logging **
```

### Additional context

I also added logs to confirm the receiver wasn't making it out of the `Start` method, caused by the ready channel not being closed.

```diff
diff --git a/receiver/kafkareceiver/kafka_receiver.go b/receiver/kafkareceiver/kafka_receiver.go
index 5f76e5295a..c943d5647f 100644
--- a/receiver/kafkareceiver/kafka_receiver.go
+++ b/receiver/kafkareceiver/kafka_receiver.go
@@ -235,7 +235,9 @@ func (c *kafkaMetricsConsumer) Start(_ context.Context, host component.Host) err
 			host.ReportFatalError(err)
 		}
 	}()
+	c.settings.Logger.Info("Waiting for consumer group to start")
 	<-metricsConsumerGroup.ready
+	c.settings.Logger.Info("Consumer group started")
 	return nil
 }
 
 ```

**Comment by djaglowski** [comment]
I agree this is a bug. If possible during `Start`, a receiver may detect unrecoverable problems and return an error. This will kill the collector. However, if a determination cannot be made quickly, or if the error may be recoverable, it should return nil and attempt to reconnect later. In this case, I think the ideal would be to have a timeout on the connection attempt and also have the [`consumeLoop`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/635ec0bb1ab82d73394023a0e565bc73c9177da8/receiver/kafkareceiver/kafka_receiver.go#L149) reconnect if necessary.

**Comment by antonblock** [comment]
Following up on our discussion at the Collector SIG, I've written a quick unit test to demonstrate this behavior. Please let me know if it's more convenient for me to put this in a draft PR, otherwise the steps below will provide a way to replicate the behavior I was seeing. With the new test added, running this from the `kafkareceiver` directory will show the error logs as the receiver fails to connect:

```fish
go test -v -timeout 30s -run ^TestMetricsReceiverStartHangs\$
```

```diff
diff --git a/receiver/kafkareceiver/kafka_receiver_test.go b/receiver/kafkareceiver/kafka_receiver_test.go
index 2497c436fa..96e87d55e4 100644
--- a/receiver/kafkareceiver/kafka_receiver_test.go
+++ b/receiver/kafkareceiver/kafka_receiver_test.go
@@ -367,6 +367,31 @@ func TestMetricsReceiverStartConsume(t *testing.T) {
 	assert.EqualError(t, err, context.Canceled.Error())
 }
 
+// TestMetricsReceiverStartHangs demonstrates that the receiver won't return from Start
+// if unable to connect to any brokers
+func TestMetricsReceiverStartHangs(t *testing.T) {
+	c, err := newMetricsReceiver(
+		Config{
+			// Ensure Kakfa isn't running at localhost:9092
+			Brokers:       []string{"localhost:9092"},
+			Encoding:      "otlp_proto",
+			InitialOffset: "latest",
+			AutoCommit: AutoCommit{
+				Interval: 1 * time.Second,
+			},
+			ClientID: "otel-test",
+			GroupID:  "otel-test",
+			Topic:    "otlp_metrics",
+		},
+		receivertest.NewNopCreateSettings(), defaultMetricsUnmarshalers(),
+		consumertest.NewNop())
+	require.NoError(t, err)
+	c.settings.Logger, _ = zap.NewDevelopment()
+
+	require.NoError(t, c.Start(context.Background(), componenttest.NewNopHost()))
+	require.NoError(t, c.Shutdown(context.Background()))
+}
+
 func TestMetricsReceiver_error(t *testing.T) {
 	zcore, logObserver := observer.New(zapcore.ErrorLevel)
 	logger := zap.New(zcore)
```

**Comment by VihasMakwana** [comment]
@antonblock for sure a bug, but if you change `metadata.full` to `true` and then try running the collector, it would fail on startup. 
If we set it to `true`, our consumer group, which is initialized in `newMetricsReceiver()`, will try to fetch all the required metadata. And that would fail as the broker is unreachable and hence, it will exit straightaway.
If we set it to `false` the consumer group won't fetch the metadata initially and it create the receiver successfully. 
This would lead to a failure in the later stage of `consumerLoop()`
